src/node_contextify.cc

```diff
@@ -527,10 +527,24 @@ void ContextifyContext::PropertySetterCallback(
       !is_function)
     return;
 
-  USE(ctx->sandbox()->Set(context, property, value));
-  if (is_contextual_store || is_function) {
-    args.GetReturnValue().Set(value);
+  Local<Value> desc;
+  bool is_get_set_property = false;
+  if (is_declared_on_sandbox &&
+      ctx->sandbox()
+          ->GetOwnPropertyDescriptor(context, property)
+          .ToLocal(&desc) &&
```

Comment:
We probably should be returning early here if this `.ToLocal()` returns `false` (i.e. getting the property descriptor throws an exception), and probably also if the `HasOwnProperty` calls below return empty `Maybe`s, rather than still trying to call `sandbox()->Set()` later on

Reply:
I thought of that too but I finally concluded: calling the set will have the benefit to apply the right behaviour. But no strong opinion at all 🙂

Reply:
If I remember well I mostly preserved the call to Set as it was done all the time starting at 18.2.0. So I may have thought 'well let's not drop it from these cases for now to reduce risks of breaking things possibly not well covered'. But in the meantime as I covered the change by many extra tests (e2e), we may give it a try. _I'll not be able to do it in the coming days, but I can either attempt to do on this PR when I'll be back or on another one. Anyway, I'll add some more checks if I do add it so that we get sure that these possible edge cases also get well covered (in some scenarios I tried during my manual checks, calling or not the Set was similar)._

Reply:
> If we do want to unconditionally set sandbox[property] = value, then maybe we should do that before the rest of the checks (and also return early in case of an exception during that assignment)?

I will update the PR to follow your suggestion by moving the line:

```cpp
USE(ctx->sandbox()->Set(context, property, value));
```

Before the if statement.

> It might be a bit more intuitive to do the equivalent of

Will try to rework it asap too.

Reply:
The first one seems correct as-is (and generally, it’s a bit unfortunate that “correct” and “shorter” are usually conflicting goals when working with V8 `Maybe`s).

I think what still needs to be addressed here is:

> and also return early in case of an exception during that assignment

i.e. replace `USE(ctx->sandbox()->Set(context, property, value));` with `if (ctx->sandbox()->Set(context, property, value).IsNothing()) return;` or similar

and currently, the PR here uses `desc_obj->HasOwnProperty(context, env->get_string()).FromJust()`, which should probably also use proper error handling (i.e. `.To()` or `.FromMaybe()` instead of `.FromJust()`) since otherwise an exception here would crash the process.